### PR TITLE
HADOOP-18891 hadoop distcp needs support to filter by file/directory attribute

### PR DESCRIPTION
**Pull request by AuthurWang2009**
### Description of PR
In some circumstances, we need to filter file/directory by file/directroy. For example, we need to filter out them by file modified time, isDir attrs, etc.

So, should we introduce a new method  public boolean shouldCopy(CopyListingFileStatus fileStatus) ? 
by this approach, we can introduce a more fluent way to do things than  public abstract boolean shouldCopy(Path path).

To achieve the goal:
1、Create a method named shouldCopy(CopyListingFileStatus fileStatus) in CopyFilter abstract method, with a supportFileStatus() swtich method which return false by default.
2、For subclasses which impl the abstract class and want to use the new method, should overwrite shouldCopy(CopyListingFileStatus fileStatus) and for the same time, return supportFileStatus()  to true.
3、This change is compatible with old use case.

As a impl:
1、I first create a abstract FileStatusCopyFilter extends CopyFilter
2、then create DirCopyFilter class extends FileStatusCopyFilter
3、and , implement UniformRecordInputFormat to support DirCopyFilter

### How was this patch tested?
added unit tests

1、add distcp.filters.class=org.apache.hadoop.tools.DirCopyFilter to distcp-default.xml or set it by -Ddistcp.filters.class=org.apache.hadoop.tools.DirCopyFilter
2、then execute distcp commands

### For code changes:

1. [x] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?

- [ ] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   5m 35s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ branch-3.3.6 Compile Tests _ |
| -1 :x: |  mvninstall  |  48m  8s | [/branch-mvninstall-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/1/artifact/out/branch-mvninstall-root.txt) |  root in branch-3.3.6 failed.  |
| +1 :green_heart: |  compile  |   0m 24s |  |  branch-3.3.6 passed  |
| +1 :green_heart: |  checkstyle  |   0m 23s |  |  branch-3.3.6 passed  |
| +1 :green_heart: |  mvnsite  |   0m 28s |  |  branch-3.3.6 passed  |
| +1 :green_heart: |  javadoc  |   0m 26s |  |  branch-3.3.6 passed  |
| +1 :green_heart: |  spotbugs  |   0m 42s |  |  branch-3.3.6 passed  |
| +1 :green_heart: |  shadedclient  |  16m  9s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 24s |  |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 18s |  |  the patch passed  |
| -1 :x: |  javac  |   0m 18s | [/results-compile-javac-hadoop-tools_hadoop-distcp.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/1/artifact/out/results-compile-javac-hadoop-tools_hadoop-distcp.txt) |  hadoop-tools_hadoop-distcp generated 1 new + 9 unchanged - 0 fixed = 10 total (was 9)  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   0m 14s | [/results-checkstyle-hadoop-tools_hadoop-distcp.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/1/artifact/out/results-checkstyle-hadoop-tools_hadoop-distcp.txt) |  hadoop-tools/hadoop-distcp: The patch generated 106 new + 4 unchanged - 0 fixed = 110 total (was 4)  |
| +1 :green_heart: |  mvnsite  |   0m 21s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 16s |  |  the patch passed  |
| +1 :green_heart: |  spotbugs  |   0m 39s |  |  the patch passed  |
| +1 :green_heart: |  shadedclient  |  15m 57s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |  13m 42s |  |  hadoop-distcp in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 28s |  |  The patch does not generate ASF License warnings.  |
|  |   | 107m  3s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6058 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient spotbugs checkstyle codespell detsecrets xmllint |
| uname | Linux 2080079e4014 4.15.0-213-generic #224-Ubuntu SMP Mon Jun 19 13:30:12 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | branch-3.3.6 / e03001452263bdcd7097c2e11994ef63fe02c6a2 |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~18.04-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/1/testReport/ |
| Max. process+thread count | 552 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-distcp U: hadoop-tools/hadoop-distcp |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/1/console |
| versions | git=2.17.1 maven=3.6.0 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by AuthurWang2009** [comment]
@hadoop-yetus 
I fixed Apache Yetus(jenkins) error: javac error

mvninstall error has nothing to do with my patch, due to nodejs.version is too low, so I update it to higher version

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DirCopyFilter.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Override
+    public boolean shouldCopy(Path path) {
+        try {
+            FileSystem fs = path.getFileSystem(this.conf);
+            if (fs.getFileStatus(path).isDirectory()) {
+                return true;
+            }
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DirCopyFilter.java]: RuntimeIOException

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DirCopyFilter.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.tools;
+
+import java.io.IOException;
+import org.apache.hadoop.conf.Configuration;
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DirCopyFilter.java]: afraid you need to use the same import ordering as everything else -its a real source of merge pain and we try to stay in control

java*

non-apache.*

org.apache.*

static *

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DirCopyFilter.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DirCopyFilter extends FileStatusCopyFilter {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/DirCopyFilter.java]: needs javadocs, and know that our indentation policy is "two spaces"

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/CopyFilter.java [hunk 4]
```diff
@@ -46,6 +46,14 @@ public void initialize() {}
    */
   public abstract boolean shouldCopy(Path path);
 
+  public boolean shouldCopy(CopyListingFileStatus fileStatus){
+    return shouldCopy(fileStatus.getPath());
+  }
+
+  public boolean supportFileStatus(){
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/CopyFilter.java]: what's this for? add javadoc to explain.

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/CopyFilter.java [hunk 5]
```diff
@@ -46,6 +46,14 @@ public void initialize() {}
    */
   public abstract boolean shouldCopy(Path path);
 
+  public boolean shouldCopy(CopyListingFileStatus fileStatus){
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/CopyFilter.java]: needs javadoc

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/UniformRecordInputFormat.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+    public RecordReader<Text, CopyListingFileStatus> createRecordReader(
+            InputSplit split, TaskAttemptContext context)
+            throws IOException, InterruptedException {
+        return new SequenceFileRecordReader<Text, CopyListingFileStatus>();
+    }
+
+    private static Path getListingFilePath(Configuration configuration) {
+        String listingFilePathString =
+                configuration.get("distcp.listing.file.path", "");
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/UniformRecordInputFormat.java]: must always reference constant strings for easse of maintenance

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/UniformRecordInputFormat.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        try {
+            FileSystem fs = listingFilePath.getFileSystem(conf);
+            if (!fs.exists(listingFilePath)) {
+                throw new IllegalArgumentException("Listing file doesn't exist at: "
+                        + listingFilePath);
+            }
+            return new SequenceFile.Reader(conf,
+                    SequenceFile.Reader.file(listingFilePath));
+        } catch (IOException exception) {
+            LOG.error("Couldn't find listing file at: " + listingFilePath, exception);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/UniformRecordInputFormat.java]: use slf4j "couldn't read {}", listingFilePAth" format

using "read" over "find" as there could be other reasons for the failure (permissions etc)

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/UniformRecordInputFormat.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                : "Couldn't find listing file. Invalid input.";
+        return new Path(listingFilePathString);
+    }
+
+    private SequenceFile.Reader getListingFileReader(Configuration conf) {
+        Path listingFilePath = getListingFilePath(conf);
+
+        try {
+            FileSystem fs = listingFilePath.getFileSystem(conf);
+            if (!fs.exists(listingFilePath)) {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/UniformRecordInputFormat.java]: use `org.apache.hadoop.util.Preconditions`

```java
checkState(fs.exists(...), "listing file doesn't exist at %s", listingFilePath);
```

File: hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/UniformRecordInputFormat.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+                throw new IllegalArgumentException("Listing file doesn't exist at: "
+                        + listingFilePath);
+            }
+            return new SequenceFile.Reader(conf,
+                    SequenceFile.Reader.file(listingFilePath));
+        } catch (IOException exception) {
+            LOG.error("Couldn't find listing file at: " + listingFilePath, exception);
+            throw new IllegalArgumentException("Couldn't find listing-file at: "
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/main/java/org/apache/hadoop/tools/mapred/UniformRecordInputFormat.java]: make this an `IllegalStateException` for consistency with the checkState.

we will need tests to simulate these failures

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+import org.junit.Test;
+
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+
+
+public class TestUniformRecordInputFormat {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java]: extends AbstractHadoopTest

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+  private static final int N_FILES = 20;
+  private static final int SIZEOF_EACH_FILE = 1024;
+  private static final Random random = new Random();
+  private static int totalFileSize = 0;
+
+  private static final Credentials CREDENTIALS = new Credentials();
+
+
+  @BeforeClass
+  public static void setup() throws Exception {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java]: can you give this a name like "setupClass" as setup() is generally used for per-test-case setup

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 28s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  |  The patch appears to include 1 new or modified test files.  |
|||| _ branch-3.3.6 Compile Tests _ |
| +0 :ok: |  mvndep  |  13m 16s |  |  Maven dependency ordering for branch  |
| -1 :x: |  mvninstall  |  14m 53s | [/branch-mvninstall-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/2/artifact/out/branch-mvninstall-root.txt) |  root in branch-3.3.6 failed.  |
| -1 :x: |  compile  |   7m 23s | [/branch-compile-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/2/artifact/out/branch-compile-root.txt) |  root in branch-3.3.6 failed.  |
| +1 :green_heart: |  checkstyle  |   1m 33s |  |  branch-3.3.6 passed  |
| +1 :green_heart: |  mvnsite  |   0m 50s |  |  branch-3.3.6 passed  |
| +1 :green_heart: |  javadoc  |   0m 45s |  |  branch-3.3.6 passed  |
| +0 :ok: |  spotbugs  |   0m 27s |  |  branch/hadoop-project no spotbugs output file (spotbugsXml.xml)  |
| +1 :green_heart: |  shadedclient  |  16m  3s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 25s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   0m 36s |  |  the patch passed  |
| -1 :x: |  compile  |   7m 12s | [/patch-compile-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/2/artifact/out/patch-compile-root.txt) |  root in the patch failed.  |
| -1 :x: |  javac  |   7m 12s | [/patch-compile-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/2/artifact/out/patch-compile-root.txt) |  root in the patch failed.  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| -0 :warning: |  checkstyle  |   1m 27s | [/results-checkstyle-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/2/artifact/out/results-checkstyle-root.txt) |  root: The patch generated 106 new + 4 unchanged - 0 fixed = 110 total (was 4)  |
| +1 :green_heart: |  mvnsite  |   0m 45s |  |  the patch passed  |
| +1 :green_heart: |  javadoc  |   0m 39s |  |  the patch passed  |
| +0 :ok: |  spotbugs  |   0m 19s |  |  hadoop-project has no data from spotbugs  |
| +1 :green_heart: |  shadedclient  |  15m 59s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   0m 17s |  |  hadoop-project in the patch passed.  |
| +1 :green_heart: |  unit  |  12m 52s |  |  hadoop-distcp in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 31s |  |  The patch does not generate ASF License warnings.  |
|  |   | 102m 18s |  |  |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6058 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient codespell detsecrets xmllint spotbugs checkstyle |
| uname | Linux f3207e1b9079 4.15.0-213-generic #224-Ubuntu SMP Mon Jun 19 13:30:12 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | branch-3.3.6 / fafa3c8423f92f1e477b8a7641e0e848878c3d1c |
| Default Java | Private Build-1.8.0_362-8u372-ga~us1-0ubuntu1~18.04-b09 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/2/testReport/ |
| Max. process+thread count | 570 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-tools/hadoop-distcp U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6058/2/console |
| versions | git=2.17.1 maven=3.6.0 spotbugs=4.2.2 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+        Path sourcePath = recordReader.getCurrentValue().getPath();
+        FileSystem fs = sourcePath.getFileSystem(configuration);
+        FileStatus fileStatus [] = fs.listStatus(sourcePath);
+        if (fileStatus.length > 1) {
+          continue;
+        }
+        currentSplitSize += fileStatus[0].getLen();
+      }
+
+      Assert.assertTrue(recordCnt == recordPerMap || recordCnt == (recordPerMap + 1));
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java]: use AssertJ.assertThat(). eg.

```java
assertJ.assertThat(recordCnt)
 .decribedAs("record count")
 .isGreaterThanOrEqualTo(recordPerMap)
 .isLessThanOrEqualTo(recordPerMap +1)
```

assertj gives wonderful error messages on assertion failures.

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+      doubleCheckedTotalSize += currentSplitSize;
+    }
+
+    Assert.assertEquals(totalFileSize, doubleCheckedTotalSize);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java]: again, assertJ here and everywhere below

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+    DataOutputStream outputStream = null;
+    try {
+      fileSystem = cluster.getFileSystem();
+      outputStream = fileSystem.create(new Path(path), true, 0);
+      int size = (int) Math.ceil(fileSize + (1 - random.nextFloat()) * fileSize);
+      outputStream.write(new byte[size]);
+      return size;
+    }
+    finally {
+      IOUtils.cleanupWithLogger(null, fileSystem, outputStream);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java]: don't close the filesystem; it is needed for the other tests

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+    }
+  }
+
+  @AfterClass
+  public static void tearDown() {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java]: 1. rename tearDownClass(); move up to just below the setup method
2. wrap the shutdown with `if (cluster != null) {...}`

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+         = stubContext.getContext();
+      recordReader.initialize(split, taskAttemptContext);
+      int recordCnt = 0;
+      while (recordReader.nextKeyValue()) {
+        recordCnt++;
+        Path sourcePath = recordReader.getCurrentValue().getPath();
+        FileSystem fs = sourcePath.getFileSystem(configuration);
+        FileStatus fileStatus [] = fs.listStatus(sourcePath);
+        if (fileStatus.length > 1) {
+          continue;
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java]: add a comment to explain

File: hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+    //we are not missing anything
+    for (InputSplit split : splits) {
+      FileSplit fileSplit = (FileSplit) split;
+      long start = fileSplit.getStart();
+      Assert.assertEquals(lastEnd, start);
+      lastEnd = start + fileSplit.getLength();
+    }
+
+    //Verify there is nothing more to read from the input file
+    SequenceFile.Reader reader
```
Comment (steveloughran) [inline hadoop-tools/hadoop-distcp/src/test/java/org/apache/hadoop/tools/mapred/TestUniformRecordInputFormat.java]: or use a try-with-resources for the automatic close

**Review by steveloughran — CHANGES_REQUESTED** [review]
Hi,

Thank you for this. You should know that we are all scared of making changes to distCp and will need multiple reviewers on this. It is just a critical part of so much workflow and fairly old code.

Here is the review criteria I use to review things, afraid I am as strict on test code as production and like tests to stress of the fairly handling of everything as that's often the code path which is not tested in production.

https://github.com/steveloughran/formality/blob/master/styleguide/styleguide.md

So: embrace assertJ and add tests for the failure conditions your code looks for (files not found etc)

And we need the distcp documentation updated. thanks.

**Comment by AuthurWang2009** [comment]
I will reopen a pull request, and merge the pull request to the trunk.

**Comment by steveloughran** [comment]
thanks. link to it from here so i'll see it

**Comment by AuthurWang2009** [comment]
I made a new PR [6070](https://github.com/apache/hadoop/pull/6070) , @steveloughran

**Pull request closed without merging**